docs/tech-notes/range-merges.md

```diff
@@ -0,0 +1,259 @@
+relevant to merges; it means that the set of stores with replicas of the LHS
+exactly matches the set of stores with replicas of the RHS. For example, this
+replica set is aligned:
+
+```
+Store 1    Store 2     Store 3     Store 4
++-----+    +-----+     +-----+     +-----+
+| P Q |    | P Q |     |     |     | P Q |
++-----+    +-----+     +-----+     +-----+
+```
+
+By requiring replica set alignment, the merge operation is reduced to a
+metadata update, albiet a tricky one, as the stores that will have a copy of
+the merged range PQ already have all the constitudent data, by virtue of having
+a copy of both P and Q before the merge begins.
+
+After verifying that the merge is sensible, the coordinator transactionally
+updates the implicated range descriptors, adjusting P's range descriptor so that
+it extends to Q's end and deleting Q's range descriptor.
+
+Then, the coordinator needs to atomically move responsibility for the data in
+the RHS to the LHS. This is tricky, as the lease on the LHS may be held by a
+different store than the lease on the RHS, The coordinator notifies the RHS that
+it is about to be subsumed and it is prohibited from serving any additional read
+or write traffic. Only when the coordinator has received an acknowledgement from
+_every_ replica of the RHS, indicating that no traffic is possibly being served
+on the RHS, does the coordinator commit the merge transaction.
+
+Like with splits, the merge transaction is committed with a special "commit
+trigger" that instructs the receiving store to update its in-memory bookkeping
+to match the updates to the range descriptors in the transaction. The moment the
+merge transaction is considered committed, the merge is complete!
+
+The remaining sections cover each of these steps in more detail.
+
+## Preconditions
+
+Not any two ranges can be merged. The first and most obvious criterion is that
+the two ranges must be adjacent. Suppose a simplified cluster that has only
+three ranges, A, B, and C:
+
+```
++-----------+-----+
+|  A  |  B  |  C  |
++-----------+-----+
+```
+
+Ranges A and B can be merged, as can ranges B and C, but not ranges A and C, as
+they are not adjacent. Note that adjacent ranges are equivalently referred to
+as "neighbors", as in, range B is range A's right-hand neighbor.
+
+The second criterion is that the replica sets must be aligned. To illustrate,
+consider a four node cluster with the default 3x replication. The allocator has
+attempted to balance ranges as evenly as possible:
+
+```
+Node  1    Node  2     Node  3     Node  4
++-----+    +-----+     +-----+     +-----+
+| P Q |    |  P  |     |  Q  |     | P Q |
++-----+    +-----+     +-----+     +-----+
+```
+
+Notice how node 2 does not have a copy of Q, and node 3 does not have a copy of
+P. These replica sets are considered "misaligned." Aligning them requires
+rebalancing Q from node 3 to node 2, or rebalancing P from node 2 to node 3:
```

Comment:
I think it's worth calling out that this is done proactively by the LHS leaseholder in order to enable merges, even if the details of how are too implementation-specific.

Reply:
Yeah, definitely need to add a section on the merge queue.